File: server_py/app/services/fhir_aggregator.py
"""FHIR data aggregation and transformation utilities."""
from typing import Dict, Any, List, Optional
from datetime import datetime
from statistics import median

from app.models import (
    SourceDataset, PatientAggregate, ObservationAggregate,
    ConditionAggregate, DemographicSummary, DatasetMetadata
)


def aggregate_fhir_data(fhir_data: Dict[str, List[Dict[str, Any]]]) -> Dict[str, Any]:
    """
    Aggregate FHIR data into summarized format for AI analysis.
    Reduces data size from ~270KB to ~2-5KB.
    """
    aggregated = {}
    
    # Aggregate patient data
    if "patients" in fhir_data and fhir_data["patients"]:
        patients = fhir_data["patients"]
        
        gender_dist = {}
        ages = []
        
        for patient in patients:
            gender = patient.get("gender", "unknown")
            gender_dist[gender] = gender_dist.get(gender, 0) + 1
            
            # Calculate age from birthDate
            if birth_date := patient.get("birthDate"):
                try:
                    birth_year = datetime.fromisoformat(birth_date).year
                    age = datetime.now().year - birth_year
                    if 0 <= age < 120:
                        ages.append(age)
                except Exception:
                    pass
        
        # Age groups
        age_groups = {
            "0-18": 0,
            "19-30": 0,
            "31-50": 0,
            "51-70": 0,
            "70+": 0
        }
        
        for age in ages:
            if age <= 18:
                age_groups["0-18"] += 1
            elif age <= 30:
                age_groups["19-30"] += 1
            elif age <= 50:
                age_groups["31-50"] += 1
            elif age <= 70:
                age_groups["51-70"] += 1
            else:
                age_groups["70+"] += 1
        
        avg_age = round(sum(ages) / len(ages), 1) if ages else None
        median_age = int(median(ages)) if ages else None
        
        aggregated["patients"] = {
            "totalCount": len(patients),
            "demographics": {
                "genderDistribution": gender_dist,
                "ageGroups": age_groups,
                "averageAge": avg_age,
                "medianAge": median_age
            },
            "sampleRecords": patients[:5]
        }
    
    # Aggregate observation data
    if "observations" in fhir_data and fhir_data["observations"]:
        observations = fhir_data["observations"]
        
        category_count = {}
        code_count = {}
        
        for obs in observations:
            # Category
            category = "unknown"
            if cat_list := obs.get("category"):
                if coding := cat_list[0].get("coding"):
                    category = coding[0].get("display") or cat_list[0].get("text", "unknown")
                else:
                    category = cat_list[0].get("text", "unknown")
            
            category_count[category] = category_count.get(category, 0) + 1
            
            # Code
            if code_obj := obs.get("code"):
                if coding := code_obj.get("coding"):
                    code = coding[0].get("code")
                    display = coding[0].get("display") or code_obj.get("text")
                else:
                    code = code_obj.get("text")
                    display = code
                
                if code:
                    if code not in code_count:
                        code_count[code] = {"count": 0, "display": display}
                    code_count[code]["count"] += 1
        
        # Top 10 common tests
        common_tests = sorted(
            [{"code": code, **data} for code, data in code_count.items()],
            key=lambda x: x["count"],
            reverse=True
        )[:10]
        
        aggregated["observations"] = {
            "totalCount": len(observations),
            "byCategory": category_count,
            "commonTests": common_tests,
            "sampleRecords": observations[:5]
        }
    
    # Aggregate condition data
    if "conditions" in fhir_data and fhir_data["conditions"]:
        conditions = fhir_data["conditions"]
        
        condition_count = {}
        severity_dist = {}
        
        for cond in conditions:
            # Condition code
            if code_obj := cond.get("code"):
                if coding := code_obj.get("coding"):
                    code = coding[0].get("code")
                    display = coding[0].get("display") or code_obj.get("text")
                else:
                    code = code_obj.get("text")
                    display = code
                
                if code:
                    if code not in condition_count:
                        condition_count[code] = {"count": 0, "display": display}
                    condition_count[code]["count"] += 1
            
            # Severity
            severity = "unknown"
            if sev_obj := cond.get("severity"):
                if coding := sev_obj.get("coding"):
                    severity = coding[0].get("display", "unknown")
                else:
                    severity = sev_obj.get("text", "unknown")
            
            severity_dist[severity] = severity_dist.get(severity, 0) + 1
        
        # Top 15 conditions
        top_conditions = sorted(
            [{"code": code, **data} for code, data in condition_count.items()],
            key=lambda x: x["count"],
            reverse=True
        )[:15]
        
        aggregated["conditions"] = {
            "totalCount": len(conditions),
            "topConditions": top_conditions,
            "severityDistribution": severity_dist,
            "sampleRecords": conditions[:5]
        }
    
    return aggregated


def create_source_dataset(fhir_data: Dict[str, List[Dict[str, Any]]]) -> SourceDataset:
    """
    Create normalized source dataset for client-side filtering and interactions.
    """
    source_data = SourceDataset()
    
    # Transform patients
    if "patients" in fhir_data and fhir_data["patients"]:
        patients = []
        for patient in fhir_data["patients"]:
            birth_date_str = patient.get("birthDate")
            age = None
            age_group = None
            
            if birth_date_str:
                try:
                    birth_year = datetime.fromisoformat(birth_date_str).year
                    age = datetime.now().year - birth_year
                    
                    if age <= 18:
                        age_group = "0-18"
                    elif age <= 30:
                        age_group = "19-30"
                    elif age <= 50:
                        age_group = "31-50"
                    elif age <= 70:
                        age_group = "51-70"
                    else:
                        age_group = "70+"
                except Exception:
                    pass
            
            patients.append(PatientAggregate(
                id=patient.get("id"),
                gender=patient.get("gender"),
                age_group=age_group,
                age=age,
                birth_date=birth_date_str
            ))
        
        source_data.patients = patients
    
    # Transform observations
    if "observations" in fhir_data and fhir_data["observations"]:
        observations = []
        for obs in fhir_data["observations"]:
            patient_id = None
            if subject := obs.get("subject"):
                if ref := subject.get("reference"):
                    patient_id = ref.split("/")[-1]
            
            category = None
            if cat_list := obs.get("category"):
                if coding := cat_list[0].get("coding"):
                    category = coding[0].get("display") or cat_list[0].get("text")
                else:
                    category = cat_list[0].get("text")
            
            code = None
            display = None
            if code_obj := obs.get("code"):
                if coding := code_obj.get("coding"):
                    code = coding[0].get("code")
                    display = coding[0].get("display") or code_obj.get("text")
                else:
                    code = code_obj.get("text")
                    display = code
            
            value = None
            unit = None
            if val_qty := obs.get("valueQuantity"):
                value = val_qty.get("value")
                unit = val_qty.get("unit")
            
            observations.append(ObservationAggregate(
                id=obs.get("id"),
                patient_id=patient_id,
                category=category,
                code=code,
                display=display,
                value=value,
                unit=unit,
                date=obs.get("effectiveDateTime")
            ))
        
        source_data.observations = observations
    
    # Transform conditions
    if "conditions" in fhir_data and fhir_data["conditions"]:
        conditions = []
        for cond in fhir_data["conditions"]:
            patient_id = None
            if subject := cond.get("subject"):
                if ref := subject.get("reference"):
                    patient_id = ref.split("/")[-1]
            
            code = None
            display = None
            if code_obj := cond.get("code"):
                if coding := code_obj.get("coding"):
                    code = coding[0].get("code")
                    display = coding[0].get("display") or code_obj.get("text")
                else:
                    code = code_obj.get("text")
                    display = code
            
            severity = None
            if sev_obj := cond.get("severity"):
                if coding := sev_obj.get("coding"):
                    severity = coding[0].get("display")
                else:
                    severity = sev_obj.get("text")
            
            category = None
            if cat_list := cond.get("category"):
                if coding := cat_list[0].get("coding"):
                    category = coding[0].get("display") or cat_list[0].get("text")
                else:
                    category = cat_list[0].get("text")
            
            conditions.append(ConditionAggregate(
                id=cond.get("id"),
                patient_id=patient_id,
                code=code,
                display=display,
                severity=severity,
                category=category,
                onset_date=cond.get("onsetDateTime")
            ))
        
        source_data.conditions = conditions
    
    # Create demographic summary
    if source_data.patients:
        gender_dist = {}
        age_groups = {
            "0-18": 0,
            "19-30": 0,
            "31-50": 0,
            "51-70": 0,
            "70+": 0
        }
        ages = []
        
        for patient in source_data.patients:
            if patient.gender:
                gender_dist[patient.gender] = gender_dist.get(patient.gender, 0) + 1
            if patient.age_group:
                age_groups[patient.age_group] += 1
            if patient.age is not None:
                ages.append(patient.age)
        
        avg_age = round(sum(ages) / len(ages), 1) if ages else None
        median_age = int(median(ages)) if ages else None
        
        source_data.demographics = DemographicSummary(
            total_patients=len(source_data.patients),
            gender_distribution=gender_dist,
            age_groups=age_groups,
            average_age=avg_age,
            median_age=median_age
        )
    
    # Create metadata
    source_data.metadata = DatasetMetadata(
        generated_at=datetime.utcnow().isoformat(),
        patient_count=len(source_data.patients) if source_data.patients else 0,
        observation_count=len(source_data.observations) if source_data.observations else 0,
        condition_count=len(source_data.conditions) if source_data.conditions else 0,
        data_source="HAPI FHIR R4"
    )
    
    return source_data

File: server/fhir-aggregator.ts
import type { 
  FHIRPatient, 
  FHIRObservation, 
  FHIRCondition,
  SourceDataset,
  PatientAggregate,
  ObservationAggregate,
  ConditionAggregate,
  DemographicSummary,
  DatasetMetadata
} from '@shared/schema';

/**
 * Aggregates FHIR data into a summarized format suitable for AI analysis
 * Instead of sending 500 full patient records, we send statistics and key insights
 */

export interface AggregatedFHIRData {
  patients?: {
    totalCount: number;
    demographics: {
      genderDistribution: Record<string, number>;
      ageGroups: Record<string, number>;
      averageAge?: number;
      medianAge?: number;
    };
    sampleRecords: any[]; // A small sample for context
  };
  observations?: {
    totalCount: number;
    byCategory: Record<string, number>;
    commonTests: Array<{ code: string; count: number; display?: string }>;
    sampleRecords: any[];
  };
  conditions?: {
    totalCount: number;
    topConditions: Array<{ code: string; count: number; display?: string }>;
    severityDistribution: Record<string, number>;
    sampleRecords: any[];
  };
}

export function aggregateFHIRData(fhirData: {
  patients?: FHIRPatient[];
  observations?: FHIRObservation[];
  conditions?: FHIRCondition[];
}): AggregatedFHIRData {
  const aggregated: AggregatedFHIRData = {};

  // Aggregate patient data
  if (fhirData.patients && fhirData.patients.length > 0) {
    const patients = fhirData.patients;
    
    // Gender distribution
    const genderDist: Record<string, number> = {};
    const ages: number[] = [];
    
    patients.forEach((patient: any) => {
      const gender = patient.gender || 'unknown';
      genderDist[gender] = (genderDist[gender] || 0) + 1;
      
      // Calculate age from birthDate
      if (patient.birthDate) {
        const birthDate = new Date(patient.birthDate);
        const age = new Date().getFullYear() - birthDate.getFullYear();
        if (age >= 0 && age < 120) {
          ages.push(age);
        }
      }
    });

    // Age groups
    const ageGroups: Record<string, number> = {
      '0-18': 0,
      '19-30': 0,
      '31-50': 0,
      '51-70': 0,
      '70+': 0
    };
    
    ages.forEach(age => {
      if (age <= 18) ageGroups['0-18']++;
      else if (age <= 30) ageGroups['19-30']++;
      else if (age <= 50) ageGroups['31-50']++;
      else if (age <= 70) ageGroups['51-70']++;
      else ageGroups['70+']++;
    });

    // Calculate average and median age
    const avgAge = ages.length > 0 ? ages.reduce((a, b) => a + b, 0) / ages.length : undefined;
    const sortedAges = [...ages].sort((a, b) => a - b);
    const medianAge = sortedAges.length > 0 
      ? sortedAges[Math.floor(sortedAges.length / 2)] 
      : undefined;

    aggregated.patients = {
      totalCount: patients.length,
      demographics: {
        genderDistribution: genderDist,
        ageGroups,
        averageAge: avgAge ? Math.round(avgAge * 10) / 10 : undefined,
        medianAge
      },
      // Include 5 sample records for context
      sampleRecords: patients.slice(0, 5)
    };
  }

  // Aggregate observation data
  if (fhirData.observations && fhirData.observations.length > 0) {
    const observations = fhirData.observations;
    
    const categoryCount: Record<string, number> = {};
    const codeCount: Map<string, { count: number; display?: string }> = new Map();
    
    observations.forEach((obs: any) => {
      const category = obs.category?.[0]?.coding?.[0]?.display || obs.category?.[0]?.text || 'unknown';
      categoryCount[category] = (categoryCount[category] || 0) + 1;
      
      const code = obs.code?.coding?.[0]?.code || obs.code?.text;
      const display = obs.code?.coding?.[0]?.display || obs.code?.text;
      if (code) {
        const existing = codeCount.get(code) || { count: 0, display };
        codeCount.set(code, { count: existing.count + 1, display });
      }
    });

    // Get top 10 most common tests
    const commonTests = Array.from(codeCount.entries())
      .sort((a, b) => b[1].count - a[1].count)
      .slice(0, 10)
      .map(([code, data]) => ({ code, ...data }));

    aggregated.observations = {
      totalCount: observations.length,
      byCategory: categoryCount,
      commonTests,
      sampleRecords: observations.slice(0, 5)
    };
  }

  // Aggregate condition data
  if (fhirData.conditions && fhirData.conditions.length > 0) {
    const conditions = fhirData.conditions;
    
    const conditionCount: Map<string, { count: number; display?: string }> = new Map();
    const severityDist: Record<string, number> = {};
    
    conditions.forEach((cond: any) => {
      const code = cond.code?.coding?.[0]?.code || cond.code?.text;
      const display = cond.code?.coding?.[0]?.display || cond.code?.text;
      if (code) {
        const existing = conditionCount.get(code) || { count: 0, display };
        conditionCount.set(code, { count: existing.count + 1, display });
      }
      
      const severity = cond.severity?.coding?.[0]?.display || cond.severity?.text || 'unknown';
      severityDist[severity] = (severityDist[severity] || 0) + 1;
    });

    // Get top 15 most common conditions
    const topConditions = Array.from(conditionCount.entries())
      .sort((a, b) => b[1].count - a[1].count)
      .slice(0, 15)
      .map(([code, data]) => ({ code, ...data }));

    aggregated.conditions = {
      totalCount: conditions.length,
      topConditions,
      severityDistribution: severityDist,
      sampleRecords: conditions.slice(0, 5)
    };
  }

  return aggregated;
}

/**
 * Creates a normalized source dataset for client-side filtering and interactions
 * This provides the raw data in a structured format that can be filtered dynamically
 */
export function createSourceDataset(fhirData: {
  patients?: FHIRPatient[];
  observations?: FHIRObservation[];
  conditions?: FHIRCondition[];
}): SourceDataset {
  const sourceData: SourceDataset = {};

  // Transform patients into normalized format
  if (fhirData.patients && fhirData.patients.length > 0) {
    sourceData.patients = fhirData.patients.map((patient: any): PatientAggregate => {
      const birthDate = patient.birthDate ? new Date(patient.birthDate) : null;
      const age = birthDate ? new Date().getFullYear() - birthDate.getFullYear() : undefined;
      
      let ageGroup: string | undefined;
      if (age !== undefined) {
        if (age <= 18) ageGroup = '0-18';
        else if (age <= 30) ageGroup = '19-30';
        else if (age <= 50) ageGroup = '31-50';
        else if (age <= 70) ageGroup = '51-70';
        else ageGroup = '70+';
      }

      return {
        id: patient.id,
        gender: patient.gender,
        ageGroup,
        age,
        birthDate: patient.birthDate
      };
    });
  }

  // Transform observations into normalized format
  if (fhirData.observations && fhirData.observations.length > 0) {
    sourceData.observations = fhirData.observations.map((obs: any): ObservationAggregate => {
      return {
        id: obs.id,
        patientId: obs.subject?.reference?.split('/')?.[1],
        category: obs.category?.[0]?.coding?.[0]?.display || obs.category?.[0]?.text,
        code: obs.code?.coding?.[0]?.code || obs.code?.text,
        display: obs.code?.coding?.[0]?.display || obs.code?.text,
        value: obs.valueQuantity?.value,
        unit: obs.valueQuantity?.unit,
        date: obs.effectiveDateTime
      };
    });
  }

  // Transform conditions into normalized format
  if (fhirData.conditions && fhirData.conditions.length > 0) {
    sourceData.conditions = fhirData.conditions.map((cond: any): ConditionAggregate => {
      return {
        id: cond.id,
        patientId: cond.subject?.reference?.split('/')?.[1],
        code: cond.code?.coding?.[0]?.code || cond.code?.text,
        display: cond.code?.coding?.[0]?.display || cond.code?.text,
        severity: cond.severity?.coding?.[0]?.display || cond.severity?.text,
        category: cond.category?.[0]?.coding?.[0]?.display || cond.category?.[0]?.text,
        onsetDate: cond.onsetDateTime
      };
    });
  }

  // Create demographic summary
  if (sourceData.patients && sourceData.patients.length > 0) {
    const genderDist: Record<string, number> = {};
    const ageGroups: Record<string, number> = {
      '0-18': 0,
      '19-30': 0,
      '31-50': 0,
      '51-70': 0,
      '70+': 0
    };
    const ages: number[] = [];

    sourceData.patients.forEach(patient => {
      if (patient.gender) {
        genderDist[patient.gender] = (genderDist[patient.gender] || 0) + 1;
      }
      if (patient.ageGroup) {
        ageGroups[patient.ageGroup]++;
      }
      if (patient.age !== undefined) {
        ages.push(patient.age);
      }
    });

    const avgAge = ages.length > 0 ? ages.reduce((a, b) => a + b, 0) / ages.length : undefined;
    const sortedAges = [...ages].sort((a, b) => a - b);
    const medianAge = sortedAges.length > 0 
      ? sortedAges[Math.floor(sortedAges.length / 2)] 
      : undefined;

    sourceData.demographics = {
      totalPatients: sourceData.patients.length,
      genderDistribution: genderDist,
      ageGroups,
      averageAge: avgAge ? Math.round(avgAge * 10) / 10 : undefined,
      medianAge
    };
  }

  // Create metadata
  sourceData.metadata = {
    generatedAt: new Date().toISOString(),
    patientCount: sourceData.patients?.length || 0,
    observationCount: sourceData.observations?.length || 0,
    conditionCount: sourceData.conditions?.length || 0,
    dataSource: 'HAPI FHIR R4'
  };

  return sourceData;
}
